File: src/EvaluateReversePolishNotation.java
// Time complexity:  O(n)
// Space Complexity: O(n)

class Solution {

    public int evalRPN(String[] tokens) {

        Set<String> operators = new HashSet<>() {{
            add("+");
            add("-");
            add("*");
            add("/");
        }};
        Stack<Integer> stack = new Stack<>();
        
        for(String token : tokens) {
            if(operators.contains(token)) {
                int secondOperand = stack.pop();
                int firstOperand = stack.pop();
                int val = 0;
                switch(token) {
                    case "+": val = firstOperand+secondOperand; break;
                    case "-": val = firstOperand-secondOperand; break;
                    case "*": val = firstOperand*secondOperand; break;
                    case "/": val = firstOperand/secondOperand; break;
                }
                stack.push(val);
            }
            else {
                stack.push(Integer.parseInt(token));
            }
        }

        return stack.pop();

    }

}

File: evaluate_reverse_polish_notation.py
# Time Complexity : O(N) where N is length of the string to be evaluated
# Space Complexity : O(N) where N is length of the string to be evaluated
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        
        for token in tokens:
            if token == "+":
                num1 = stack.pop()
                num2 = stack.pop()
                stack.append(num2 + num1)
            elif token == "-":
                num1 = stack.pop()
                num2 = stack.pop()
                stack.append(num2 - num1)
            elif token == "*":
                num1 = stack.pop()
                num2 = stack.pop()
                stack.append(num2 * num1)
            elif token == "/":
                num1 = stack.pop()
                num2 = stack.pop()
                stack.append(int(num2 / num1))
            else:
                stack.append(int(token))
           
        return stack.pop()
                